onnxruntime/python/onnxruntime_pybind_state.cc

```diff
@@ -257,11 +259,17 @@ void CustomOpLibrary::UnloadLibrary() {
 #endif  // !defined(ORT_MINIMAL_BUILD)
 
 template <typename T>
-void AddNonTensor(const OrtValue& val, std::vector<py::object>& pyobjs, const DataTransferManager* /*data_transfer_manager*/) {
+void AddNonTensor(const OrtValue& val, std::vector<py::object>& pyobjs,
+                  const DataTransferManager* /*data_transfer_manager*/,
+                  const std::unordered_map<OrtDevice::DeviceType, MemCpyFunc>* /*mem_cpy_to_host_functions*/) {
   pyobjs.push_back(py::cast(val.Get<T>()));
 }
 
-void GetPyObjFromTensor(const Tensor& rtensor, py::object& obj, const DataTransferManager* data_transfer_manager) {
+// In all cases, we may not have access to a DataTransferManager, hence the user may specify functions that
+// pretty much does what a DataTransferManager does - copy data from device(s) to the host
+void GetPyObjFromTensor(const Tensor& rtensor, py::object& obj,
+                        const DataTransferManager* data_transfer_manager,
+                        const std::unordered_map<OrtDevice::DeviceType, MemCpyFunc>* mem_cpy_to_host_functions) {
```

Comment:
DataTransferManager is coupled too tightly with an InferenceSession. For cases where we need to create a PyObject from a tensor stand-alone (without the context of a tensor) - users can feed in a map of device to host memcpy functions. Strings are not supported - they never were.

Reply:
Would it be simpler if we require a session to be present to use this functionality? 

What's the use case where you want to copy to host and there isn't a session? I would have expected you were copying outputs from running the model.

---
In reply to: [492464929](https://github.com/microsoft/onnxruntime/pull/5248#discussion_r492464929) [](ancestors = 492464929)

Reply:
The use case is if someone wants to create a numpy object from an OrtValue (ortvalue.numpy()) whose memory is on the device - we would have to copy to the host in this case and we want to be able to do this without the context of a session. 

---
In reply to: [497066823](https://github.com/microsoft/onnxruntime/pull/5248#discussion_r497066823) [](ancestors = 497066823,492464929)

Reply:
we can but I think we will be duplicating a lot of code in the function body and a lot of code is common except the way in which the device to host copies are handled. This approach doesn't duplicate code. Are you concerned about users passing in both ? There is a comment which says the DataTransferManager is preferred in case both are passed in...

---
In reply to: [493127114](https://github.com/microsoft/onnxruntime/pull/5248#discussion_r493127114) [](ancestors = 493127114)